### PR TITLE
Add delete user fuctionality, increase verbosity and general refactoring

### PR DESCRIPTION
**Pull request by mira-miracoli**
Important info:
`delete` a user means:
The user will notified using the notification framwork and an "urgent" notification, which should be delivered by email and can not be unsubscribed. The user will be marked as deleted in the database, all running jobs will be cancelled, all sessions will be terminated and login is blocked.
However it will NOT delete all their data, histories etc. on the spot.
This only happens when a user is `purged` 

Delete user functionality works as follows:
The user can specify the flag `--delete-user <MIN_SEVERITY>` , where `<MIN_SEVERITY>` must be one of the defined `VALID_SEVERITIES` (low, medium and high), this is not case sensitive.

Everytime the script matches a file in a jwd to the malware database, 
and if the severity level of that malware is EQUAL OR HIGHER than the specified  `<MIN_SEVERITY>`, the Galaxy user ID is added to a set named `delete_users`.

Once the iteration over the running jobs found in the databse is done, the script iterates over the `delete_users` set, requests an encoded version of their user ID from Galaxy's API (`GALAXY_BASE_URL` and `GALAXY_API_KEY` must be set in the `~/.bashrc` file of the user running WALL·E) and then creates a notification that will stay in the database for 6 months and sends a `DELETE` request to `/api/users/<encoded_id>`.
- [x] Once there is a notification framework with a mandatory email channel, the script shall send a message to that user before the deletion happens. (see https://github.com/usegalaxy-eu/issues/issues/514#issuecomment-1959158284)

The second, more minor change in this PR is the amount of information that is logged in verbosity mode:
~~~diff
- TIMESTAMP GALAXY_USER JOB_ID MALWARE_CLASS MALWARE MALWARE_VERSION PATH
+ TIMESTAMP MALWARE_SEVERITY USER_ID USER_NAME USER_MAIL TOOL_ID GALAXY_JOB_ID RUNNER_JOB_ID RUNNER_NAME MALWARE_CLASS OBJECT_STORE_ID MALWARE_VERSION PATH
~~~

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -28,11 +28,17 @@
...
-    - 'export PGPASSFILE="{{ walle_pgpass_file }}"'
+    regexp: "^export {{ item.key }}="
+    line: 'export {{ item.key }}="{{ item.value}}"'
+  with_items: "{{ walle_envs_database }}"
 
+- name: Add env variables for user deletion (WallE)
+  ansible.builtin.lineinfile:
+    path: "{{ walle_bashrc }}"
+    regexp: "^export {{ item.key }}="
+    line: 'export {{ item.key }}="{{ item.value}}"'
```
Comment (bgruening) [inline tasks/main.yml:39]:
```suggestion
    line: 'export {{ item.key }}="{{ item.value }}"'
```

File: tasks/main.yml [hunk 2]
```diff
@@ -28,11 +28,17 @@
 - name: Add env variable for database (WallE)
   ansible.builtin.lineinfile:
     path: "{{ walle_bashrc }}"
-    line: "{{ item }}"
-  with_items:
-    - 'export MALWARE_LIB="{{ walle_malware_database_location }}/{{ walle_database_file }}"'
-    - 'export PGPASSFILE="{{ walle_pgpass_file }}"'
+    regexp: "^export {{ item.key }}="
+    line: 'export {{ item.key }}="{{ item.value}}"'
```
Comment (bgruening) [inline tasks/main.yml:32]:
```suggestion
    line: 'export {{ item.key }}="{{ item.value }}"'
```

File: templates/walle.py [hunk 3]
```diff
@@ -131,28 +175,43 @@ def make_parser() -> argparse.ArgumentParser:
...
     )
-
+    my_parser.add_argument(
+        "--delete-user",
+        metavar="MIN_SEVERITY",
+        choices=VALID_SEVERITIES,
+        type=convert_str_to_severity,
+        help="Delete user when severity level is equal or higher. \
+            Make sure that you know what the consequences are on your \
+            instance, when a user is set to deleted. \
```
Comment (bgruening) [inline templates/walle.py:185]:
```suggestion
            instance, when a user is set to deleted (e.g. when a user is purged automatically after deletion). \
```

File: templates/walle.py [hunk 4]
```diff
@@ -422,40 +444,113 @@ def get_running_jobs(self, tool=None) -> [Job]:
...
+                    runner_name=runner_name,
+                    object_store_id=object_store_id,
                 )
             )
         return running_jobs_list
 
 
+def get_path_from_env_or_error(env: str) -> pathlib.Path:
+    if os.environ.get(env):
+        if (path := pathlib.Path(os.environ.get(env).strip())).exists():
```
Comment (bgruening) [inline templates/walle.py:473]: oh the walrus in wild :)

File: templates/walle.py [hunk 5]
```diff
@@ -422,40 +444,113 @@ def get_running_jobs(self, tool=None) -> [Job]:
...
+    )
     jobs = db.get_running_jobs(args.tool)
+    if args.delete_user:
+        api = GalaxyAPI(
+            api_key=get_str_from_env_or_error("GALAXY_API_KEY"),
+            base_url=get_str_from_env_or_error("GALAXY_BASE_URL"),
+        )
+        delete_users = set()
+    if not args.verbose:
+        report_users = set()
```
Comment (bgruening) [inline templates/walle.py:547]: I would not put the variables in the if condition, just fill the set and only filter print() with arg.verbose

File: templates/walle.py (hunk 4 above)
Comment (mira-miracoli) [reply templates/walle.py:473]: Pretty common in go 🐹

File: templates/walle.py (hunk 5 above)
Comment (mira-miracoli) [reply templates/walle.py]: like this?

**Comment by mira-miracoli** [comment]
shall I split it up in multiple PRs so we can merge the fix for symlinks and increased verbosity now

**Comment by mira-miracoli** [comment]
I think this is finally ready for review :)

**Review by sanjaysrikakulam — COMMENTED** [review]
It reads fine at first glance. Has this been tested? Would it be possible to add some mock tests?

**Comment by mira-miracoli** [comment]
Yes, I tested on my instance and I can give you access, so you can try yourself if you want.
Test coverage would be nice, yes, I can start working on this today

**Comment by sanjaysrikakulam** [comment]
Oh, cool then. Yup, test coverage would be excellent! Thank you!

**Comment by mira-miracoli** [comment]
But it might be inefficient, because I never wrote cli tests for python before and the last unit test is 5 y ago. also if we merge #7 we might need to rewrite/change it again.

**Review by sanjaysrikakulam — APPROVED** [review]

**Comment by sanjaysrikakulam** [comment]
Feel free to merge or wait for @bgruening's review.

**Comment by bgruening** [comment]
Go for it! Fingers crossed!

**Comment by mira-miracoli** [comment]
Merging will not change our production servers – [versions are fixed](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/c131f14c76801f2be0e27745c22cf084722d68e1/requirements.yaml#L130C1-L131C19)

**Pull request merged**